### PR TITLE
Patterns: Add user pattern categories to post editor inserter patterns tab

### PR DESCRIPTION
**Pull request by glendaviesnz**
## What?
Add user pattern categories to the post editor block inserter panel

## Why?
With the addition of the ability for users to categorize their patterns they will expect them to be ordered  by those categories in the inserter.

## How?
Used the new `__experimentalBlockPatternCategories` block editor setting

This PR was part of a group of PRs that were merged back into https://github.com/WordPress/gutenberg/pull/53835 before being merged into trunk.

A full list of the PRs involved in this work can be [found here](https://github.com/WordPress/gutenberg/issues/53164).

## Testing Instructions

- Add some unsynced patterns with categories assigned (currently on unsynced patterns display in the patterns category panel - follow up PRs will merge synced patterns into this)
- Click on the inserter Patterns panel and make sure the patterns appear on the correct category and can be inserted
- Also add an uncategorized pattern and make sure it appears in an `Uncategorized` category

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/f7dd06b6-a677-4117-9f90-6e6f97bce0aa

### CODE REVIEW

File: packages/block-editor/src/store/selectors.js [hunk 1]
```diff
@@ -2293,7 +2293,8 @@ const checkAllowListRecursive = ( blocks, allowedBlockTypes ) => {
 function getUnsyncedPatterns( state ) {
 	const reusableBlocks =
 		state?.settings?.__experimentalReusableBlocks ?? EMPTY_ARRAY;
-
+	const { patternCatogoriesMap: categories } =
```
Comment (aaronrobertshaw) [inline packages/block-editor/src/store/selectors.js]:
```suggestion
	const { patternCategoriesMap: categories } =
```

Typo. Which if fixed on the primary PR will need updating here.

**Review by aaronrobertshaw — APPROVED** [review]
This is testing as advertised for me. 

✅ Unsynced patterns display in appropriate user-created categories
✅ Uncategorized patterns show correctly in `Uncategorized`

Just one small typo I left a comment on.

**Review by aaronrobertshaw — COMMENTED** [review]
I've given this another test and still appears to be working as advertised for me. Checked with TT3 & TT2 after it was mentioned the pattern lists showed different patterns between this and trunk.

One question though, was there an explicit decision on whether user-created patterns should appear before theme patterns or not? 

Currently, after creating a custom pattern it is displayed at the very bottom of the pattern inserter list. Could that be easier for a user to miss?

**Comment by glendaviesnz** [comment]
> One question though, was there an explicit decision on whether user-created patterns should appear before theme patterns or not? Currently, after creating a custom pattern it is displayed at the very bottom of the pattern inserter list. Could that be easier for a user to miss?

I have switched it so user patterns appear first - @SaxonF do you have a view on whether this is the best approach?

**Comment by SaxonF** [comment]
@glendaviesnz makes sense to me

**Comment by glendaviesnz** [comment]
This was merged with some failing tests, but is going into a feature branch which needs rebasing to get the failing tests resolved, so failing tests will be sorted on that branch prior to merging to trunk.

**Pull request merged**